keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
+
+Users can create dual stack services and control which up gets assigned to Service Primary IP as the following:
+1. Set `spec.ipFamilyPolicy: DualStack`. 
+2. If `spec.ipFamilyPolicy: nil` and `len(spec.ipFamilies) == 2` the resulting service will have a defaulted `spec.ipFamilyPolicy: DualStack`.
+3. If `spec.ipFamilies: nil` then it will be defaulted to the same order of ip families as set in `--services-cluster-ip-range`
+4. If `spec.ipFamilies: ['ip family']` it will be defaulted to `spec.ipFamilies: ['ip family', 'other ip family']`.
+5. If `spec.ipFamilies: ['ip family', 'ip family']` the order will be respected by the system and `spec.ClusterIP` will be allocated from the first entry in the array.
+
+> It is implicitly assumed that a Service with `spec.ipFamilyPolicy: nil` is always a single stack, however users can change this behavior and the order of ip families using admission controllers.
+
+The below is an example for the dual stack Service where the user only set  `spec.ipFamilyPolicy: DualStack` on a cluster that has `--services-cluster-ip-range=<ipc4 cidr>, <ipv6 cidr>`
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  ipFamilyPolicy: DualStack # set by user
+  ipFamilies: # defaulted
+    - IPv4
+    - IPv6
+  clusterIP: 1.2.3.4  # allocated
+  clusterIPs: # allocated
+    - 1.2.3.4
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+```
 
+The below Service is a dual stack service where the user is requesting a certain order of ip family assignment, irrespective of the `--services-cluster-ip-range` setting.
+ 
 ```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
-  ipFamily: IPv6
+  type: ClusterIP
+  ipFamilyPolicy: DualStack # defaulted
+  ipFamilies: # set by user
+    - IPv6
+    - IPv4
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1
+    - 1.2.3.4
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+status:
+  ipFamilies:
+    - IPv4
+    - IPv6
 ```
 
-Dual-stack services (having both an IPv4 and IPv6 cluster IP) are currently out of scope of this proposal. Should there be use-cases where dual-stack Services are needed then we can revisit.
+##### Notes on NodePort Allocation
+
+NodePort reservation will be constant across both families. That mean a reservation of `NodePort: 12345` will happen across both families irrespective of the ip family of the service.
 
 #### Type Headless services
 
-\<TBD\>
+1. For Headless with selector, the value of `spec.ipFamilyPolicy` and `spec.ipFamilies` will drive the endpoint selection. Endpoint slice controller will select endpoints with ip families that match the value of these two flags.
+2. For Headless without selector the values of `spec.ipFamilyPolicy` and `spec.ipFamilies` are treated as saved as they were provided by the user.
```

Comment:
But do they have any meaning in this case? Should we just require them to be unset?

Reply:
I think they can serve as hints to whatever is deriving Endpoints for the service.